src/Operation/Watch.php

```diff
@@ -241,40 +254,36 @@ private function createResumeCallable(Manager $manager)
                 unset($this->changeStreamOptions['startAtOperationTime']);
             }
 
+            // Select a new server using the original read preference
+            $server = $manager->selectServer($this->aggregateOptions['readPreference']);
+
             /* If we captured an operation time from the first aggregate command
              * and there is no "resumeAfter" option, set "startAtOperationTime"
              * so that we can resume from the original aggregate's time. */
-            if ($this->operationTime !== null && ! isset($this->changeStreamOptions['resumeAfter'])) {
+            if ($this->operationTime !== null && ! isset($this->changeStreamOptions['resumeAfter']) &&
+                \MongoDB\server_supports_feature($server, self::$wireVersionForStartAtOperationTime)) {
                 $this->changeStreamOptions['startAtOperationTime'] = $this->operationTime;
             }
 
+            // Recreate the aggregate command and execute to obtain a new cursor
             $this->aggregate = $this->createAggregate();
+            $cursor = $this->executeAggregate($server);
 
-            /* Select a new server using the read preference, execute this
-             * operation on it, and return the new ChangeStream. */
-            $server = $manager->selectServer($this->aggregateOptions['readPreference']);
-
-            return $this->execute($server);
+            return [$cursor, $this->isFirstBatchIsEmpty];
         };
     }
 
     /**
-     * Execute the aggregate command and optionally capture its operation time.
+     * Execute the aggregate command.
+     *
+     * The command will be executed using APM so that we can capture its
+     * operation time and/or firstBatch size.
      *
      * @param Server $server
      * @return Cursor
      */
     private function executeAggregate(Server $server)
     {
-        /* If we've already captured an operation time or the server does not
-         * support resuming from an operation time (e.g. MongoDB 3.6), execute
-         * the aggregation directly and return its cursor. */
-        if ($this->operationTime !== null || ! \MongoDB\server_supports_feature($server, self::$wireVersionForStartAtOperationTime)) {
```

Comment:
Why was this removed?

Reply:
The original logic was actually a bit backwards, but it worked in practice.

The server has reported operation times for as long as change streams have been supported (since 3.6); however, the `startAtOperationTime` option has only been supported since 4.0. Previously, `createResumeCallable` would decide to use `startAtOperationTime` purely by whether `$this->operationTime` was set, as we would only ever set that if this condition was true and we ran the `aggregate` command through APM.

With this PR, we need to always check whether `aggregate`'s `firstBatch` is empty and therefore always run through APM. That means that `$this->operationTime` was always captured and could result in the driver attempting to use `startAtOperationTime` to resume against a 3.6 server (in the absence of a resume token).

To address this, I moved the `server_supports_feature` check to `createResumeCallable` before we decide to use that option. That fixes the issue with resuming against a 3.6 server when no changes have been returned. This also means that the APM callback can continue collecting `$this->operationTime` all the time. Alternatively, I suppose we could have added a `server_supports_feature` within the APM callback and decide whether to ignore the operation time, but `createResumeCallable` seems more straightforward since that's when we actually attempt to specify the option.

Reply:
That makes sense, thanks!